Refactor connect context

Lbqds opened the pull request:
Split the `AlephiumConnectContext` into two smaller contexts to avoid rerendering

h0ngcha0 commented on packages/web3-react/src/hooks/useConnect.tsx:
```diff
@@ -96,27 +96,29 @@ export function useConnect(options: ConnectOptions) {
       ?.enable({
         ...options,
-        onDisconnected: disconnectAlephium
+        onDisconnected: () => {
```
Question: What's the reason to set `onDisconnected` to no-op?

Lbqds replied:
Because we have already called `context.setSignerProvider(undefined)` and `context.setAccount(undefined)` at here: https://github.com/alephium/alephium-web3/blob/7cdf1fde136b81a459564c1c4b0285af86c4b96a/packages/web3-react/src/hooks/useConnect.tsx#L103

When the user clicks the disconnect button, the function call sequence is: `disconnectAlephium` -> `alephium.disconnect()` -> `onDisconnected`, which will cause `disconnectAlephium` to be executed twice. If I understand correctly, we only need to perform cleanup in either the `onDisconnect` callback or `disconnectAlephium`.

Lbqds replied:
Great point 👍, I just tested it, and the context will also be cleaned up in this case.

h0ngcha0 commented on packages/web3-react/src/components/AlephiumConnect.tsx:
```diff
@@ -86,17 +80,38 @@ export const AlephiumConnectProvider: React.FC<AlephiumConnectProviderProps> = (
   )
 }
+
+export const AlephiumConnectProvider: React.FC<{ children?: React.ReactNode }> = ({ children }) => {
```
IIUC, now we need to wrap the app with both `AlephiumConnectProvider` and `ConnectSettingProvider`. Maybe we should update [README](https://github.com/alephium/alephium-web3/blob/master/packages/web3-react/README.md?plain=1#L38). I will update docs as well.

h0ngcha0 replied:
Sorry, just saw in the next commit we have included `ConnectSettingProvider` in `AlephiumConnectProvider`.

Lbqds replied:
Will update the readme, thanks!

h0ngcha0 commented:
A general question:
Previously we have a single context, right now we split them into two and created `AlephiumWalletProvider` to wrap them together.  When will the previous case cause re-render while the new approach doesn't?

Lbqds commented:
> A general question: Previously we have a single context, right now we split them into two and created `AlephiumWalletProvider` to wrap them together. When will the previous case cause re-render while the new approach doesn't?

According to the [react doc](https://react.dev/reference/react/useContext#optimizing-re-renders-when-passing-objects-and-functions), when the value within the context updates, all components using the context will be rerendered. 

When we click the wallet connect button, both the `open` and `route` properties within the context will update, and then leading to multiple renders.

However, we have now split the context into two parts. Users only need to call `useAlephiumConnectContext` to retrieve wallet connection information, so updates to `ConnectSettingContext` will not trigger rerenders of the user's components. And his approach is inspired by `connectkit` and `wagmi`.

Lbqds commented on packages/web3/src/api/utils.ts:
```diff
@@ -26,3 +27,20 @@ export function convertHttpResponse<T>(response: { status: number; data: T; erro
+    const tokens1Size = tokens1?.length ?? 0
+    if (tokens0Size !== tokens1Size) return false
+    if (tokens0Size === 0) return true
+    return tokens0!.every((t0) => tokens1!.find((t1) => t0.id === t1.id && t0.amount === t1.amount) !== undefined)
```
Fixed, thanks!

polarker reviewed (approved):
LGTM

h0ngcha0 reviewed (approved):
LGMT :+1:
